docs/design/proposals/healer-controller.md

```diff
@@ -0,0 +1,120 @@
+# Introduction
+There is [external health monitoring](https://github.com/kubernetes-csi/external-health-monitor/) which does mostly similar stuff, but actions are not in the scope of it, all it can do is log the health condition of the mounts.
+
+# Volume Healer Controller
+Before we begin it is noteworthy to know that the existing healer logic remains the same.
+
+**Now the obvious question we expect is "what do we achieve with the new controller?"**
+* Moving the k8s specific code out of ceph-csi, to make ceph-csi orchestrator independent.
+* Achieve a way to monitor the health of NBD volumes.
+* We currently do not have a way to act on failed/crashed NBD volumes with the existing healer logic. Once the controller detects that the volume is not healthy, it will have an opportunity to act on it.
+* This will be a generic framework for any userspace mounters added in the future, example ceph-fuse
+
+# Details
+
+                    .-------------------.
+                    |     CSI-Addons    |
+                    | Healer Controller |
+                    '-------------------'
+                              |
+                              | gRPC
+                              |
+            .---------+------------------------------.
+            |         |                              |
+            |  .------------.        .------------.  |
+            |  | CSI-Addons |  gRPC  |    CSI     |  |
+            |  |  side-car  |--------| NodePlugin |  |
+            |  '------------'        '------------'  |
+            | CSI-driver Pod                         |
+            '----------------------------------------'
+
+## CSI-Addons Healer Controller
+* The healer controller will be implemented as part of the CSI-Addons controller driver. It is the one who triggers controller RPC to check the health condition of the CSI volumes.
+* Watch for the `CSIAddonsNode` object, if the object is created/modified by the side-car do below operations
```

Comment:
why `CSIAddonsNode` and not `VolumeAttachment`? ideally use the same  formatting/capitalization for all Kuberenets objects

Reply:
`VolumeAttachment` is specific to a single volume, which is not our interest here. We should be watching for events that are at node level i.e. `CSIAddonsNode`

Volume level abnormalities should be taken care by the periodic healing checks.

Reply:
Ah! I realized what you meant now.